llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/Vectorization.cpp

```diff
@@ -35,74 +37,128 @@ SmallVector<int64_t> interchange(ArrayRef<int64_t> elements,
   return vec;
 }
 
-struct PackOpVectorizationPattern : OpRewritePattern<PackOp> {
+Value getInputOrPaddedInput(OpBuilder &builder, PackOp packOp) {
+  Value input = packOp.getInput();
+  if (!packOp.getPaddingValue()) {
+    return input;
+  }
+
+  Location loc = packOp.getLoc();
+  SmallVector<OpFoldResult> lowPadding, highPadding;
+  auto zeroAttr = builder.getIndexAttr(0);
+  ShapedType inputType = packOp.getInputType();
+  int64_t inputRank = inputType.getRank();
+  lowPadding.append(inputRank, zeroAttr);
+
+  SmallVector<int64_t> paddedShape;
+  DenseMap<int64_t, OpFoldResult> tileAndPosMapping =
+      packOp.getDimAndTileMapping();
+  for (int64_t dim = 0; dim < inputRank; ++dim) {
+    int64_t size = inputType.getDimSize(dim);
+    if (!tileAndPosMapping.count(dim)) {
+      paddedShape.push_back(size);
+      highPadding.push_back(zeroAttr);
+      continue;
+    }
+
+    Optional<int64_t> tileSize =
+        getConstantIntValue(tileAndPosMapping.lookup(dim));
+    assert(!inputType.isDynamicDim(dim) && tileSize.hasValue() &&
+           "something goes really wrong...");
+    int64_t sizeWithPad = llvm::alignTo(size, tileSize.getValue());
+    highPadding.push_back(builder.getIndexAttr(sizeWithPad - size));
+    paddedShape.push_back(sizeWithPad);
+  }
+  auto resultType =
+      RankedTensorType::get(paddedShape, inputType.getElementType());
+  return tensor::createPadScalarOp(resultType, input, packOp.getPaddingValue(),
+                                   lowPadding, highPadding,
+                                   /*nofold=*/false, loc, builder);
+}
+
+// Creates a linalg.generic that transposes input using permutation indices.
+// Example: (M1, M0, N1, N0) -> (M1, N1, M0, N0) if indices = {0, 2, 1, 3}.
+Value createTransposeOp(Location loc, OpBuilder &builder, Value input,
```

Comment:
note: a linalg.transpose up is coming upstream

Reply:
Is it more than a convenience method to create transpose?